habitat-lab/habitat/gui/text_drawer.py

```diff
@@ -0,0 +1,94 @@
+
+DEFAULT_FONT_PATH = "./fonts/ProggyClean.ttf"
+
+# the maximum number of chars displayable in the app window
+# using the magnum text module
+MAX_DISPLAY_TEXT_CHARS = 1024
+
+# how much to displace window text relative to the center of the
+# app window (e.g if you want the display text in the top left of
+# the app window, you will displace the text
+# window width * -TEXT_DELTA_FROM_CENTER in the x axis and
+# window height * TEXT_DELTA_FROM_CENTER in the y axis, as the text
+# position defaults to the middle of the app window)
+TEXT_DELTA_FROM_CENTER = 0.49
+
+# font size of the magnum in-window display text
+DISPLAY_FONT_SIZE = 16.0
+
+
+class TextDrawer:
+    def __init__(
+        self,
+        framebuffer_size: mn.Vector2i,
+        relative_path_to_font: str = DEFAULT_FONT_PATH,
+        max_display_text_chars: int = MAX_DISPLAY_TEXT_CHARS,
+        text_delta_from_center: float = TEXT_DELTA_FROM_CENTER,
+        display_font_size: float = DISPLAY_FONT_SIZE,
+    ) -> None:
+        self._framebuffer_size = framebuffer_size
+
+        # Load a TrueTypeFont plugin and open the font file
+        self._display_font = text.FontManager().load_and_instantiate(
+            "TrueTypeFont"
+        )
+        self._display_font.open_file(
+            os.path.join(os.path.dirname(__file__), relative_path_to_font),
+            13,
+        )
+
+        # Glyphs we need to render everything
+        self._glyph_cache = text.GlyphCache(mn.Vector2i(256))
+        self._display_font.fill_glyph_cache(
+            self._glyph_cache,
+            string.ascii_lowercase
+            + string.ascii_uppercase
+            + string.digits
+            + ":-_+,.! %µ",
+        )
+
+        # magnum text object that displays CPU/GPU usage data in the app window
+        self._window_text = text.Renderer2D(
+            self._display_font,
+            self._glyph_cache,
+            display_font_size,
+            text.Alignment.TOP_LEFT,
+        )
+        self._window_text.reserve(max_display_text_chars)
+
+        # text object transform in window space is Projection matrix times Translation Matrix
+        # put text in top left of window
+        self._window_text_transform = mn.Matrix3.projection(
+            self._framebuffer_size
+        ) @ mn.Matrix3.translation(
+            mn.Vector2(self._framebuffer_size)
+            * mn.Vector2(
+                -text_delta_from_center,
+                text_delta_from_center,
+            )
+        )
+        self._shader = shaders.VectorGL2D()
+
+        # make magnum text background transparent
+        mn.gl.Renderer.enable(mn.gl.Renderer.Feature.BLENDING)
+        mn.gl.Renderer.set_blend_function(
+            mn.gl.Renderer.BlendFunction.ONE,
+            mn.gl.Renderer.BlendFunction.ONE_MINUS_SOURCE_ALPHA,
+        )
+        mn.gl.Renderer.set_blend_equation(
+            mn.gl.Renderer.BlendEquation.ADD, mn.gl.Renderer.BlendEquation.ADD
+        )
+
+    def draw_text(self, text_to_draw):
```

Comment:
Can you add documentation for how to use this function? E.g. what is text_to_draw, how do I specify where text gets printed on-screen, etc.

Reply:
I can think of two possible interfaces that would be nice to expose to user code (Drivers):
(1) `draw_text(str, on_screen_position)`, where `on_screen_position` does something similar to CSS to help position the text, e.g. a combination of left|right and top|bottom alignment plus a pixel offset.
(2) `add_text_line(str)`, where we just let the user code provide multiple lines of text, and then the GuiRenderer decides how to arrange them on-screen.

We want user code (Drivers) to be able to call `draw_text/add_text_line` from anywhere in sim code, at any time. The TextDrawer needs to save these requests internally and then "flush" them later (actually drawing to OpenGL). Again, this is exactly the paradigm used by DebugLineRender.

Open to other ideas here on the interface for specifying/arranging the text!

Reply:
And just a heads up that someday in the future, we will want 3D-positioned text, e.g. `draw_text(str, world_position)`, which will use the current camera and project to screen space to come up with the 2D screen position.

Reply:
Done! I hope my implementation looks good to you @eundersander.
I've combined your suggestions `draw_text(str, on_screen_position)`  and `add_text_line(str)` into a single function with kwargs [add_text](https://github.com/facebookresearch/habitat-lab/blob/09a3274e79c9d6f8a534f68f21111e22f804c407/habitat-lab/habitat/gui/text_drawer.py#L84). If no kwargs are set it is equivalent to `add_text_line(str)` , but with kwargs it can be configured like `draw_text(str, on_screen_position)`.